### DISCUSSION TITLE
Unclear how I can reliably set the Partition Key in DynamoDB in my EF repository layer

### DISCUSSION THREAD

**Discussion by davidbrinton** [thepirat000/Audit.NET#613] [Q&A]
I'm using Audit.Net for all of our audits and have it turned on to audit every create or update event using the Entity Framework interceptor. We're pushing our events to DynamoDB. Up until now we've set the Partition Key to a random UUID for every event written, but this is not a great option as we can't look up all changes to a particular object. For example, if we want to see all of the events against the Orders table for order with OrderId 123. For this reason we're wanting to change the Partition Key to always be the primary key value of the table we're auditing.

It's not clear to me how I can do this in a thread safe way that supports all tables. I think I can do something like this in a basic operation

```
public async Task UpdateNoteAsync(Note note)
{
    using (AuditScope scope = AuditScope.Create(new AuditScopeOptions {  ExtraFields = new Dictionary<string, object> { { "EntityId", noteId } }}))
    using (NoteContext context = _ContextFactory.GetContext<NoteContext>())
    {
        // Update the note here

        await context.SaveChangesAsync();
    }
}
```

I think that will cause it to write the `noteId` value into the `EntityId` field and I think that would be thread safe. Is that right? If so, then, my next wonder would be what would happen if I am updating more than one table in a single block like that? For instance if I'm updating the `Note` table and the `Order` table? How would the scope know when to put `NoteId` in the `EntityId` field and when to put `OrderId` in the `EntityId` field?

I'm thinking there must be some way to achieve what I'm trying to achieve more easily though. For example, if there was a global configuration I could do like this?

```
Configuration.Setup()
    .UseDynamoDB(config => config
    .UseUrl(auditingDatabaseConnection)
    .Table(AUDITING_DATABASE_TABLE_NAME)
    .SetAttribute(AUDITING_DATABASE_ATTRIBUTE_NAME, ev => Guid.NewGuid()))
    .TableSettings(new List<TableSetting>
    { 
        new TableSetting { TableName = "Order", SourceFieldName = "OrderId", OutputFieldName = "EntityId", OutputFieldPrefix = "order_" },
        new TableSetting { TableName = "Note", SourceFieldName = "NoteId", OutputFieldName = "EntityId", OutputFieldPrefix = "note_" }
    });
```

Thanks for the help on learning what my options are here.

**Comment by thepirat000** [comment]
If you added the EF interceptor and surrounded the save changes with an `AuditScope.Create`, you'll end up having 2 events logged. 
Is that what you want?

If so, you'll need something to co-relate the events. For example, if you're in an ASP.NET Core application, you could use the request trace identifier. I think the easiest way to attach the TraceIdentifier to the event is via a Custom Action in your startup logic:

```c#
public void ConfigureServices(IServiceCollection services)
{
    services.AddHttpContextAccessor();
    ...
}

public void Configure(IApplicationBuilder app, IHttpContextAccessor contextAccessor)
{
    Audit.Core.Configuration.AddCustomAction(ActionType.OnScopeCreated, scope =>
    {
        var httpContext = contextAccessor.HttpContext;
        scope.Event.CustomFields["CorrelationId"] = httpContext.TraceIdentifier;
    });
    ...
}
```

Now all your events will have the "CorrelationId" custom field set to a unique identifier by request

**Reply by davidbrinton** [reply]
Oh, gosh, no I didn't want to log two events. I guess I didn't think through how that would work. I just want to log all of my CRUD events in my database to a table in DynamoDB and I want the Partition Key for the DynamoDB to be the name of the table, plus the primary key so that I can easily do look ups on the audit logs in the DynamoDB table. For example

```
SELECT * FROM AuditTable WHERE EntityId = 'order_1234'
```

In this case EntityId is my Partition Key and this query would return all audits logged for the order table for order 1234.

**Reply by davidbrinton** [reply]
@thepirat000 This is what I have currently and it's just putting a GUID as the EventId, which makes it basically impossible to find anything if I need to.

```
Configuration.Setup()
    .UseDynamoDB(config => config
    .UseUrl(auditingDatabaseConnection)
    .Table(AUDITING_DATABASE_TABLE_NAME)
    .SetAttribute(AUDITING_DATABASE_ATTRIBUTE_NAME, ev => Guid.NewGuid()));
```

**Reply by thepirat000** [reply]
Oh, I see.

So, assuming that:
- All your SaveChanges calls affect one and only one entity alone
- All the entities audited have a single column primary key

Then, you can set the attribute name dynamically from the audit event:

```c#
Audit.Core.Configuration.Setup()
    .UseDynamoDB(config => config
        .UseUrl(auditingDatabaseConnection)
        .Table(AUDITING_DATABASE_TABLE_NAME)
        .SetAttribute(AUDITING_DATABASE_ATTRIBUTE_NAME, ev =>
        {
            var firstEntry = ev.GetEntityFrameworkEvent()?.Entries.FirstOrDefault();
            var table = firstEntry?.Table;
            var id = firstEntry?.PrimaryKey.FirstOrDefault().Value.ToString();

            return $"{table}_{id}";
        }));
```

But usually, a SaveChanges call affects multiple tables, so I'm not sure if that will work for you. 

If you happen to have a SaveChanges that insert into 2 tables, then the audit event including both inserts, will have the partition as the first table:id that was inserted.

**Reply by davidbrinton** [reply]
@thepirat000 Yeah, we do often affect more than one table in the update. I could rewrite it a bit to look at all entries individually, but it's still only returning a single value even though Audit.Net is writing more than one audit entry.

Is there a way for me to set an attribute for each audit entry that's about to be written?

**Reply by thepirat000** [reply]
If you affect more than one entity/table in the same savechanges, do you want a separate item in dynamo for each?

**Reply by davidbrinton** [reply]
@thepirat000 Yes, that would be the goal. We're getting that right now with the Audit.Net interceptor and we don't want to change that, we just need a more search-friendly PartitionKey value, something that if someone from the business asks us to find all of the audit events for a specific order, we can get that quickly and easily. The last time they needed something we had to dump the whole table and it took a couple of hours. My plan was to do tableName_tableKey as the value for the PartitionKey, but if there is another solution that is easier to implement I would be all ears.

BTW, thanks for helping so quickly and frequently.

**Reply by thepirat000** [reply]
If you're using Audit.EF for the audit event generation and Audit.NET.DybamoDb to store the events, then you'll get 1 dynamo document per each SaveChanges,  even if it affected more than one table. Each document will contain a list of the entities affected.

I'm not sure why you say you're getting that right now with audit.net

**Reply by davidbrinton** [reply]
@thepirat000 Oh, really? OK, I guess I just assumed I was getting one row for each affected table. Is there a way to do that then?

**Reply by thepirat000** [reply]
Yes, the only data provider that does that by default is the Entity Framework provider (which stores the audit events using EF)...

Take into account that the SaveChanges interceptor is just that, an interceptor of the SaveChanges call, there is no way in EF to intercept by table.

So Audit.EF will generate one audit event that will potentially be stored as multiple documents in your dynamo db.

I think the only way to do that is to create a custom data provider, which shouldn't be too much difficult.

For example:

```c#
public class MyCustomDynamoDataProvider : AuditDataProvider
{
    private const string AUDITING_DATABASE_TABLE_NAME = "TABLE";
    private const string AUDITING_DATABASE_ATTRIBUTE_NAME = "PART_KEY";
    private readonly AmazonDynamoDBClient _client;
    
    public MyCustomDynamoDataProvider(string auditingDatabaseConnection)
    {
        _client = new AmazonDynamoDBClient(new AmazonDynamoDBConfig() { ServiceURL = auditingDatabaseConnection });
    }

    public override object InsertEvent(AuditEvent auditEvent) => InsertEventAsync(auditEvent).GetAwaiter().GetResult();

    public override async Task<object> InsertEventAsync(AuditEvent auditEvent, CancellationToken cancellationToken = default)
    {
        var efEvent = auditEvent.GetEntityFrameworkEvent();

        var dynamoTable = Table.LoadTable(_client, AUDITING_DATABASE_TABLE_NAME);
        var batch = dynamoTable.CreateBatchWrite();

        // For each entity modified
        foreach (var entry in efEvent.Entries)
        {
            // Set its partition key field
            entry.CustomFields[AUDITING_DATABASE_ATTRIBUTE_NAME] = $"{entry.Table}_{entry.PrimaryKey.First().Value}";

            // Also add any fields you need from the root AuditEvent, for example:
            entry.CustomFields["Environment"] = auditEvent.Environment;
            entry.CustomFields["StartDate"] = auditEvent.StartDate;
            entry.CustomFields["Duration"] = auditEvent.Duration; // the duration is the total duration of the SaveChanges call, not just this entity

            // Add the document to the batch 
            batch.AddDocumentToPut(Document.FromJson(entry.ToJson()));  // Note the document is created from this particular entry and not the entire auditevent
        }

        // Insert all the documents at once
        await batch.ExecuteAsync(cancellationToken);

        return new object();
    }
}
``` 

And you just configure it like this:

```c#
Audit.Core.Configuration.Setup()
    .Use(new MyCustomDynamoDataProvider("THE URL"));
```

I didn't test this at all, but should give you an idea of how can you save multiple documents for a single audit event.

**Reply by davidbrinton** [reply]
@thepirat000 Wow, thank you. I will look at implementing this today and will come back and mark this as the answer if I don't run into anything unforeseen.

**Reply by davidbrinton** [reply]
@thepirat000 OK, I've started implementing this and I have two questions.

I don't see how to tie this into Entity Framework since adding a row to a table in the database won't trigger an event. I know I can't use the interceptor since that only works on `SaveChanges`. I've looked at the documentation once again and it looks like there is an Entity Framework Data Provider which looks like it expects to store the audit events using Entity Framework, which I don't think supports DynamoDB and there is also an `Audit.EntityFramework` extension which contains the interceptor I've been using, but also a low-level interceptor that looks promising but I guess I'm still unsure as to how to connect this custom AuditDataProvider to events that occur in Entity Framework.

The other question I had was, as this is an `InsertEventAsync` override, does that mean it only works for row creation statements, or can I also use this for `UPDATE` and `DELETE` operations?

**Reply by thepirat000** [reply]
The interceptors are only aware of things that happen within your code using an intercepted EF DbContext, they will not intercept operations that happen externally, like manually adding a row to a table, etc. If you need to audit any operation at the server level, you should consider using a native tool for your database, like [SQL Server Audit Components](https://learn.microsoft.com/en-us/sql/relational-databases/security/auditing/sql-server-audit-database-engine?view=sql-server-ver16).

Also, this is what connects the data provider to all the events that happen within Audit.NET:

```c#
Audit.Core.Configuration.Setup()
    .Use(new MyCustomDynamoDataProvider("THE URL"));
```

Here you're settling the global data provider that will handle all the audit events.

Note the `InsertEventAsync` in the data provider refers to the Audit Event insertion, not the operation in the database. Each EF audit event will contain 1 or more operations to the database. So, if you're only creating EF audit events, you can just override `InsertEvent / InsertEventAsync` like in my previous example. 
> The Data provider also has `ReplaceEvent / ReplaceEventAsync` methods, but you don't need to override those for the EF events, since they are never replaced.

**Reply by davidbrinton** [reply]
@thepirat000 OK, I think I understand, this `AuditDataProvider` customizes how events are mapped to the DynamoDB. So, can I still use the interceptor to trigger the event and this provider will split the different updates out into multiple rows? I think that's what you're saying, is that right?

**Reply by thepirat000** [reply]
That's correct!
Sorry if I wasn't clear, my english's not the best

**Reply by davidbrinton** [reply]
@thepirat000 I think your English is quite clear, I just was missing an understanding about how some of the underlying functionality in Audit.Net worked.

OK, I'm going to give this a shot. Thank you.

**Comment by thepirat000** [comment]
Another option is to use the [DbContext.ContextId](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.contextid?view=efcore-7.0) as a custom field to correlate the events, you'll need to change the order of the `usings`:

```c#
using (NoteContext context = _ContextFactory.GetContext<NoteContext>())
using (AuditScope scope = AuditScope.Create(new AuditScopeOptions { ExtraFields = new { EntityId = noteId, ContextId = context.ContextId.ToString() } }))
{
   ...
```

So now the manual event will also include the ContextId property (EF events already includes the context id by default) which is unique per DbContext instance
